libbeat/licenser/manager.go

```diff
@@ -0,0 +1,315 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package licenser
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+// OSSLicense default license to use.
+var (
+	OSSLicense = &License{
+		UUID:   uuid.NewV4(),
+		Type:   OSS,
+		Mode:   OSS,
+		Status: Active,
+		Features: features{
+			Graph:      graph{},
+			Logstash:   logstash{},
+			ML:         ml{},
+			Monitoring: monitoring{},
+			Rollup:     rollup{},
+			Security:   security{},
+			Watcher:    watcher{},
+		},
+	}
+)
+
+// Watcher allows a type to receive a new event when a new license is received.
+type Watcher interface {
+	OnNewLicense(license License)
+	OnManagerStopped()
+}
+
+// Fetcher interface implements the mechanism to retrieve a License. Currently we only
+// support license coming from the '/_xpack' rest api.
+type Fetcher interface {
+	Fetch() (*License, error)
+}
+
+// Errors returned by the manager.
+var (
+	ErrWatcherAlreadyExist = errors.New("watcher already exist")
+	ErrWatcherDoesntExist  = errors.New("watcher doesn't exist")
+
+	ErrManagerStopped = errors.New("license manager is stopped")
+	ErrNoLicenseFound = errors.New("no license found")
+)
+
+// Backoff values when the remote cluster is not responding.
+var (
+	maxBackoff  = time.Duration(60)
+	initBackoff = time.Duration(5)
+)
+
+// Manager keeps tracks of license management, it uses a fetcher usually the ElasticFetcher to
+// retrieve a licence from a specific cluster.
+//
+// Starting the manager will starts a go routine to periodically query the license fetcher.
+// if an error occur on the fetcher we will retries until we successfully
+// receive a new license. During that period we start a grace counter, we assume the license is
+// still valid during the grace period, when this period expire we will keep retrying but the previous
+// license will be invalidated and we will fallback to the OSS license.
+//
+// Retrieving the current license:
+// - Call the `Get()` on the manager instance.
+// - Or register a `Watcher` with the manager to receive the new license and acts on it, you will
+// also receive an event when the Manager is stopped.
+//
+//
+// Notes:
+// - When the manager is started no license is set by default.
+// - When a license is invalidated, we fallback to the OSS License and the watchers get notified.
```

Comment:
Will monitoring continue to function if the OSS license is active? In case of monitoring we don't check the license, but the availability of said feature in ES.

Reply:
In the current implementation the OSS has all the feature disabled, so when the license is invalidated monitoring will also stop. But this should be transient since the manager will still continue to try to get the new license from the cluster and the monitoring should resume.

When this case is happening, I expect that monitoring will not be able to reach the cluster to send monitoring events.